test_runner/README.md

```diff
@@ -62,36 +62,67 @@ Exit after the first test failure:
 `pytest -x ...`
 (there are many more pytest options; run `pytest -h` to see them.)
 
+### Writing a test
 
-### Building new tests
+Every test needs a Zenith Environment, or ZenithEnv to operate in. A Zenith Environment
+is like a little cloud-in-a-box, and consists of a Pageserver, 0-N Safekeepers, and
+compute Postgres nodes. The connections between them can be configured to use JWT
+authentication tokens, and some other configuration options can be tweaked too.
 
-The tests make heavy use of pytest fixtures. You can read about how they work here: https://docs.pytest.org/en/stable/fixture.html
+The easiest way to get access to a Zenith Environment is by using the 'zenith_simple_env'
+fixture. The 'simple' env may be shared across multiple tests, so don't shut down the nodes
+or make other destructive changes in that environment.
 
-Essentially, this means that each time you see a fixture named as an input parameter, the function with that name will be run and passed as a parameter to the function.
+For more complicated cases, you can build a custom Zenith Environment, with the 'zenith_env'
+fixture:
 
-So this code:
-
-```python
-def test_something(zenith_cli, pg_bin):
-    pass
 ```
+test_foobar(zenith_env: ZenithEnv):
+    # Prescribe the environment.
+    env = zenith_env
+
+	# We want to have 3 safekeeper nodes
+	env.set_num_safekeepers(3)
 
-... will run the fixtures called `zenith_cli` and `pg_bin` and deliver those results to the test function.
+    # Use JWT authentication in the connections to the page server
+    env.set_pageserver_auth(True)
 
-Fixtures can't be imported using the normal python syntax. Instead, use this:
+    # Now create the environment. This initializes the repository, starts
+	# up the page server, and the safekeepers
+	env.init()
```

Comment:
Maybe it's better if `zenith_env` is a builder/factory for the `env` variable or uses [markers](https://docs.pytest.org/en/6.2.x/fixture.html#using-markers-to-pass-data-to-fixtures) to clearly separate configuration/initialization on a type level?

Reply:
>  Rename the fixture to zenith_env_builder. It's rarely used, so it's good to avoid confusion with shorter env.

Done.

> Pass parameters directly to init instead of messing with the state. As far as I understand, you have a clear distinction between arguments unlikely to change in an individual test (repo_dir, port_distributor) and configurable ones.

Hmm, I'm not sure about that. I'm envisioning that we will have many more options in the future, so the list of parameters could grow large. Maybe that's OK if you use named parameter syntax. But I left it as it is for now.

> Consider using __call__ instead of init. Harder to grep, but the call syntax looks cooler. No other serious advantages :)

Heh, didnt' know about that syntax :-).